### ISSUE TITLE
Latest VDU driver changes (Sep 24) - video broken on Apple IIGS

### ISSUE THREAD

**Issue by bobbimanners** [bobbimanners/Applecorn#102]
Testing on my ROM01 GS:

- None of the text only modes is working (`MODE 6` gives a bank screen now)
- 80 column `MODE 3` seems confused.  I will post a screenshot.
- `MODE 2` bitmapped mode is still working, mostly.

**Comment by bobbimanners** [comment]
It is entirely possible I broke this somehow in my merge.

**Comment by bobbimanners** [comment]
I tested your version too on real GS and none of the modes is working for some reason.  Will investigate.

**Comment by bobbimanners** [comment]
If I comment out the `LDA #$E0` for the GS arround line 606, then the 40 col modes work again.

So the new GS stuff doesn't work on real hardware at the moment.  None of the GS emulators is that brilliant unfortunately, so stuff sometimes works in the emulator but not on real iron.

**Comment by bobbimanners** [comment]
Some good news though.  On the GS, testing with your build, the 80 column screen is half working (odd columns) even with the Alternate Text Mode CDA disabled.  So your long writes are working for E0 I think ... maybe not for E1?

**Comment by bobbimanners** [comment]
I took a first look at this.  First thing is to set Merlin-16 to build 65816 code (so we don't have to mess with `DB` for the opcode).  Two `XC` pseudo instructions in a row set the assembler to 65816 mode.

Second thing ... the 65816 is running in 6502 emulation mode.  One needs to do this:
```
             CLC                     ; Enter 65816 native mode
             XCE
             SEP   #$30              ; 8-bit M and 
```
to switch to native mode and to set the accumulator and index regs to 8 bit.  And ...
```
            SEC
            XCE
```
to go back to emulation mode.

**Comment by bobbimanners** [comment]
So to put it all together we end up with something like:
```
:L5          CLC                     ; Enter 65816 native mode
             XCE
             SEP   #$30              ; 8-bit M and X
             LDA   [VDUADDR],Y
             STA   [VDUADDR2],Y
             SEC                     ; Enter emulation mode
             XCE
```

**Comment by bobbimanners** [comment]
With that fix, odd columns work.  I think the remaining issue is you are always using bank $E0 but the other half of the 80 column mode is in $E1.  Unless I am missing something.

**Comment by bobbimanners** [comment]
Finally gotten the sodding thing working.  More or less.

**Comment by bobbimanners** [comment]
OK, here is what I discovered and it ain't pretty.

We want to write to the screen at E00800-E00BFF and E10800-E10BFF.  However it turns out that the 'WRITEMAIN' and 'WRITEAUX' softswitches seem to override the bank selection so if 'WRITEAUX' is enabled only one of the banks actually gets updated.  I have fixed this by shoving in a `>>> WRTMAIN` before the write to $0E/$E1 and `>>> WRTAUX` after.

However, the same thing seems to happen on reading also, so we have problems with scrolling and copy cursor.  Maybe these need to be done from main memory.  What a nuisance this is!

**Comment by bobbimanners** [comment]
** Reverted the PR for now **

**Comment by jgharston** [comment]
From the 65816 documentation I've been using, the [longzp],Y addressing mode should be available when running as a 6502, you shouldn't need to switch to 65816 mode.

It works on KEGS, but that's no guarrantee!

**Comment by bobbimanners** [comment]
Do you have a reference to that documentation by any chance?  I am using a doc from WDC called 'Programming tge 65816.pdf'  [Attached].

[Programming the 65816.pdf](https://github.com/bobbimanners/Applecorn/files/7231708/Programming.the.65816.pdf)

**Comment by jgharston** [comment]
Yep, that's the one.
Starting with the lastest vdu.s I've got most of the updates merged, with the AppleGS-specific code temporarily removed for the mo. I've put the kettle on to finish off the integration with gfx.s. First fix is that CLS moves to (0,0).  :)

**Comment by bobbimanners** [comment]
Awesome.  I am around this afternoon for testing and merging :)

**Comment by bobbimanners** [comment]
Apple IIgs stuff can wait to the side until we have an elegant way to do it.

**Comment by bobbimanners** [comment]
So I just took the dogs out for a walk and had a good discussion with them about the Apple IIgs video.

Here is their suggestion ... we do the same thing we did for the MODE 2 graphics screen.  When we write chars to banks E0/E1 (which I already have working) we also keep on writing them to 00/01 (aka main and aux).  That way the copy cursor keeps working with no changes and we can read from main/aux when scrolling and don't have to worry about reading from E0/E1.  

Should be much simpler!

While I have your attention ... should we make the HGR mode `MODE 1` instead of 2 since it is 40 cols?

**Comment by jgharston** [comment]
Getting there. Changed to MODE 1, as it is MODE 1. Putting the kettle on for a chunk of testing. I'll be uploading updated vdu.s and gfx.s later - plus a *one* *line* change to mainmem.gfx.

**Comment by bobbimanners** [comment]
Sounds good.  Standing by here :)
I can do some testing if you like!

**Comment by jgharston** [comment]
The one line change to mainmem.gfx is:
* On entry: char is in A
DRAWCHAR    >>>   ENTMAIN
*            AND   #$7F ; Don't!
;)

**Comment by bobbimanners** [comment]
I think I did that cos my `FONT.DAT` only has 128 chars.  I can build that file in various ways (see the script in the `font` subdir.)

**Comment by jgharston** [comment]
VDU and HGR updates merged. Text modes and Graphics modes (ie MODE 1) nicely working.
All 65816 code disabled for the mo. Have been testing on Apple2ix and KEGS on Android and AppleWin and KEGS32 on Windows.

**Comment by jgharston** [comment]
I think the data in the Turner font file needs to be inverted, as the MouseChar characters are all inverted. If the whole file is inverted, then the inverted and non-inverted normal characters just swap with each other, and the MouseChar characters become non-inverted.

**Comment by bobbimanners** [comment]
Your latest code seems to work with the unmolested font file, so I am using that for now.

**Comment by bobbimanners** [comment]
@jgharston seems you are right about being able to use the new 65816 instructions even in emulation mode!  I didn't know about that ... I stand corrected!!

https://softpixel.com/~cwright/sianse/docs/65816NFO.HTM

**Comment by bobbimanners** [comment]
Apple IIGS video is working now (with some quirks but it is basiclly there.)

I will clean it up and check it in.

**Comment by bobbimanners** [comment]
@jgharston 
I tweaked the Apple IIgs code with the insertation of some switches to main memory (`>>> WRTMAIN` / `>>> WRTAUX`).  With those changes, it seems to be mostly working on my ROM1 GS.  I have checked in those changes but left them disabled for now.

Things that do not work in GS mode:
- Copy editor, especially in MODE1 bitmapped mode. 
- The character under the cursor is not preserved properly.

However 80 column MODE 3 is working nicely, including scrolling and CLS.  It is definitely getting there!!

So writing to banks E0 and E1 is working.  Maybe reading from them is going wrong though.
